### PR TITLE
Generic passthrough handler should be able to cover all Kanto functio…

### PR DESCRIPTION
**Pull request by hristobojilov**
[#21] Generic passthrough handler should be able to cover all Kanto functionalities

Signed-off-by: Hristo Bozhilov <Hristo.Bozhilov@bosch.io>

### CODE REVIEW

File: routing/message/handlers/passthrough/telemetry_handler.go [hunk 1]
```diff
@@ -15,15 +15,18 @@ package passthrough
...
 
+	"github.com/eclipse-kanto/azure-connector/config"
+	"github.com/eclipse-kanto/azure-connector/routing"
 	"github.com/eclipse-kanto/azure-connector/routing/message/handlers"
 )
 
 const (
 	telemetryHandlerName = "passthrough_telemetry_handler"
+
+	topicsEvent = "event/#,e/#,telemetry/#,t/#,command//+/res/#,c//+/s/#"
```
Comment (e-grigorov) [inline routing/message/handlers/passthrough/telemetry_handler.go:29]: `event/#` - by definition the topics should be in the format: event//${device-id} i.e. the filter should be: `event//+` Do we need to have such relaxed matching?

File: routing/message/handlers/passthrough/telemetry_handler.go [hunk 2]
```diff
@@ -33,9 +36,9 @@ type telemetryHandler struct {
 }
 
 // CreateTelemetryHandler instantiates a new passthrough telemetry handler that forward messages received from local message broker on the given topics as device-to-cloud messages to Azure IoT Hub.
```
Comment (e-grigorov) [inline routing/message/handlers/passthrough/telemetry_handler.go:38]: The doc has to be aligned with the current implementation.

File: routing/topics.go [hunk 3]
```diff
@@ -26,6 +28,9 @@ const (
 
 	remoteCloudTopicFmt     = "devices/%s/messages/devicebound/#"
 	remoteTelemetryTopicFmt = "devices/%s/messages/events/%s"
+
+	localCmdTopicLongFmt  = "command//%s:%s/req//%s"
```
Comment (e-grigorov) [inline routing/topics.go:32]: I am not sure that I can understand completely the format. What is the representation of the request id?

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: routing/message/handlers/passthrough/command_handler.go [hunk 4]
```diff
@@ -34,10 +34,9 @@ const (
 	msgInvalidCloudCommand = "invalid cloud command"
 )
 
-// A simple command passthrough handler that forwards all cloud-to-device messages from Azure IoT Hub to local MQTT broker on a preconfigured topic.
 type commandHandler struct{}
 
-// CreateCommandHandler instantiates a new command handler that forward cloud-to-device messages to the local message broker using the given topic.
+// CreateCommandHandler instantiates a new command handler that forwards cloud-to-device messages to the local message broker using topics command//+/req/# and c//+/q/#.
```
Comment (e-grigorov) [inline routing/message/handlers/passthrough/command_handler.go:39]: topics command//+/req/# and c//+/q/# - these two are not topics but filters. In general, I would suggest to rephrase it to: to the local message broker as Hono commands.

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: routing/topics.go [hunk 5]
```diff
@@ -43,3 +48,23 @@ func CreateTelemetryTopic(deviceID, msgID string) string {
...
+		env.Topic.Namespace,
+		env.Topic.EntityName,
+		env.Headers.CorrelationID(),
+		env.Topic.Action,
+	)
+}
+
+// CreateLocalCmdTopicShort constructs the local MQTT topic for receiving C2D messages from an Azure IoT Hub device.
+func CreateLocalCmdTopicShort(env *protocol.Envelope) string {
+	return fmt.Sprintf(localCmdTopicShortFmt,
```
Comment (e-grigorov) [inline routing/topics.go:64]: Clean up the duplicated code.

File: routing/topics.go [hunk 6]
```diff
@@ -51,10 +51,20 @@ func CreateTelemetryTopic(deviceID, msgID string) string {
 
 // CreateLocalCmdTopicLong constructs the local MQTT topic for receiving C2D messages from an Azure IoT Hub device.
 func CreateLocalCmdTopicLong(env *protocol.Envelope) string {
-	return fmt.Sprintf(localCmdTopicLongFmt, env.Topic.Namespace, env.Topic.EntityName, env.Topic.Action)
+	return fmt.Sprintf(localCmdTopicLongFmt,
+		env.Topic.Namespace,
+		env.Topic.EntityName,
+		env.Headers.CorrelationID(),
```
[the remark this answers is not in the document]
Comment (hristobojilov) [reply routing/topics.go:57]: In case correlation-id is missing it will be replaced by an empty string.

[the remark this answers is not in the document]
Comment (e-grigorov) [reply routing/topics.go]: Ok, perfect, then we are in case of one-way comment topic.

**Review by antonia-avramova — APPROVED** [review]

File: routing/message/handlers/passthrough/telemetry_handler.go (hunk 1 above)
Comment (e-grigorov) [reply routing/message/handlers/passthrough/telemetry_handler.go:29]: The topics are taken from suite connector. They will be kept in this way.

**Review by e-grigorov — APPROVED** [review]

**Review by stoyan-zoubev — CHANGES_REQUESTED** [review]

File: routing/message/handlers/passthrough/telemetry_handler.go [hunk 7]
```diff
@@ -15,27 +15,29 @@ package passthrough
...
-// telemetryHandler forwards incoming device messages published on certain local topics to the Azure Iot Hub.
 type telemetryHandler struct {
 	deviceID string
 	topics   string
 }
 
-// CreateTelemetryHandler instantiates a new passthrough telemetry handler that forward messages received from local message broker on the given topics as device-to-cloud messages to Azure IoT Hub.
-func CreateTelemetryHandler(topics string) handlers.TelemetryHandler {
+// CreateTelemetryHandler instantiates a new passthrough telemetry handler that forwards messages received from local message broker on event, telemetry and command response topics as device-to-cloud messages to Azure IoT Hub.
+func CreateTelemetryHandler() handlers.TelemetryHandler {
```
[the remark this answers is not in the document]
Comment (hristobojilov) [reply routing/message/handlers/passthrough/telemetry_handler.go:38]: We will leave this method as it was and add additional one CreateDefaultTelemetryHandler without topics parameter.

[the remark this answers is not in the document]
Comment (stoyan-zoubev) [reply routing/message/handlers/passthrough/telemetry_handler.go:38]: Great!

File: routing/message/handlers/passthrough/command_handler_test.go [hunk 8]
```diff
@@ -13,34 +13,82 @@
...
+	"github.com/ThreeDotsLabs/watermill/message"
+
 	"github.com/eclipse-kanto/suite-connector/connector"
 
-	"github.com/ThreeDotsLabs/watermill/message"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
 func TestCreateCommandHandler(t *testing.T) {
```
Comment (e-grigorov) [inline routing/message/handlers/passthrough/command_handler_test.go:27]: It should be: TestCreateDefaultCommandHandler

**Review by stoyan-zoubev — APPROVED** [review]

**Review by stoyan-zoubev — APPROVED** [review]

File: routing/message/handlers/passthrough/command_handler_test.go [hunk 9]
```diff
@@ -13,34 +13,82 @@
...
-	messageHandler := CreateCommandHandler("command-topic")
+	messageHandler := CreateDefaultCommandHandler()
 	require.NoError(t, messageHandler.Init(nil))
 	assert.Equal(t, commandHandlerName, messageHandler.Name())
 }
```
Comment (e-grigorov) [inline routing/message/handlers/passthrough/command_handler_test.go:32]: Do we need TestCreateCommandHandler?

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: routing/message/handlers/passthrough/command_handler_test.go (hunk 9 above)
Comment (e-grigorov) [reply routing/message/handlers/passthrough/command_handler_test.go:32]: Invalid, the command handler doesn't have two create methods.

**Review by e-grigorov — APPROVED** [review]
